src/osd/scrubber/scrub_machine.cc

```diff
@@ -654,62 +654,182 @@ ScrubMachine::~ScrubMachine() = default;
+ * as an abort request.
+ *
+ * Process:
+ * - if already reserved: clear existing reservation, then continue
+ * - ask the OSD for the "reservation resource"
+ * - if granted: mark it internally and notify the Primary.
+ * - otherwise: just notify the requesting primary.
+ */
+void ActiveAsReplica::on_reserve_req(const ReplicaReserveReq& ev)
 {
-  dout(10) << "-- state -->> ReplicaIdle" << dendl;
+  DECLARE_LOCALS;  // 'scrbr' & 'pg_id' aliases
+  dout(10) << "ActiveAsReplica::on_reserve_req()" << dendl;
+
+  if (reserved_by_my_primary) {
+    dout(10) << "ActiveAsReplica::on_reserve_req(): already reserved" << dendl;
+    // clear the existing reservation
+    clr_remote_reservation();  // clears the flag, too
+  }
+
+  // ask the OSD for the reservation
+  const auto ret = get_resource();
+  if (ret.granted) {
+    reserved_by_my_primary = true;
+    dout(10) << fmt::format("{}: reserved? yes", __func__) << dendl;
+  } else {
+    dout(10) << fmt::format("{}: reserved? no ({})", __func__, ret.error_msg)
+	     << dendl;
+  }
+
+  Message* reply = new MOSDScrubReserve(
+      spg_t(pg_id.pgid, m_pg->get_primary().shard), ev.m_op->sent_epoch, ret.op,
+      m_pg->pg_whoami);
+  m_osds->send_message_osd_cluster(reply, ev.m_op->get_req()->get_connection());
 }
 
-ReplicaIdle::~ReplicaIdle()
+
+void ActiveAsReplica::on_release(const ReplicaRelease& ev)
 {
+  DECLARE_LOCALS;  // 'scrbr' & 'pg_id' aliases
+  if (!reserved_by_my_primary) {
+    dout(5) << fmt::format(
+		   "ActiveAsReplica::on_release() from {}: not reserved!",
+		   ev.m_from)
+	    << dendl;
+    return;
+  }
+  dout(10) << fmt::format("ActiveAsReplica::on_release() from {}", ev.m_from)
+	   << dendl;
+  clr_remote_reservation();
+}
+
+
+ActiveAsReplica::ReservationAttemptRes ActiveAsReplica::get_resource()
+{
+  using ReservationAttemptRes = ActiveAsReplica::ReservationAttemptRes;
+  DECLARE_LOCALS;  // 'scrbr' & 'pg_id' aliases
+  if (!scrbr->get_pg_cct()->_conf.get_val<bool>("osd_scrub_during_recovery") &&
+      m_osds->is_recovery_active()) {
+    return ReservationAttemptRes{
+	MOSDScrubReserve::REJECT, "recovery is active", false};
+  }
+
+  if (m_osds->get_scrub_services().inc_scrubs_remote(scrbr->get_spgid().pgid)) {
+    return ReservationAttemptRes{MOSDScrubReserve::GRANT, "", true};
+  } else {
+    return ReservationAttemptRes{
+	MOSDScrubReserve::REJECT, "failed to reserve remotely", false};
+  }
 }
 
 
-// ----------------------- ReplicaActiveOp --------------------------------
+void ActiveAsReplica::clr_remote_reservation()
+{
+  DECLARE_LOCALS;  // 'scrbr' & 'pg_id' aliases
+  dout(10) << "ActiveAsReplica::clr_remote_reservation()" << dendl;
+  m_osds->get_scrub_services().dec_scrubs_remote(scrbr->get_spgid().pgid);
+  reserved_by_my_primary = false;
+}
+
+
+sc::result ActiveAsReplica::react(const IntervalChanged&)
+{
+  dout(10) << "ActiveAsReplica::react(const IntervalChanged&)" << dendl;
+  clr_remote_reservation();
```

Comment:
Do this in the exit() function or the destructor.

Reply:
Same with ReplicaActiveOp, replica_handling_done should probably be done once upon exiting the state.  I think you should only need a reaction list entry for ReplicaActive.